syntax = "proto3";

package cider.services.build.companion;

// Represents a CoG checkout on user's workstation.
message WorkspaceState {
  // Absolute path for the checkout in the workstation, including the repo_root.
  // e.g. /google/cog/cloud/$USER/$WORKSPACE/googleplex-android/
  // $WORKSPACE can be alias or id.
  string repo_dir = 1;
  // Relative to repo_dir.
  repeated string active_file_path = 2;
  // Absolute path to output directory in workstation.
  string out_dir = 3;
  // Absolute path to compile_commands.json in workstation.
  string comp_db_path = 4;
  // Absolute path to the main ninja file, e.g.
  // ../android-aosp/out/combined-aosp_arm.ninja
  string ninja_path = 5;
}

// Indicates the success/failure for analysis.
message Status {
  enum Code {
    OK = 0;
    FAILURE = 1;
  }
  Code code = 1;
  // Details about the status, might be displayed to user.
  optional string message = 2;
}

// Provides all the targets that are pre-requisities for running language
// services on active_file_paths.
message DepsRequest {
  WorkspaceState state = 1;
}
message DepsResponse {
  // Build dependencies of a source file for providing language services.
  message Deps {
    // Relative to workspace root.
    string source_file = 1;
    // Ninja target to execute for generating dep.
    repeated string build_target = 2;
    optional Status status = 3;
  }
  repeated Deps deps = 1;
  optional Status status = 2;
}

// Returns all the information necessary for providing language services for the
// active files.
message InputsRequest {
  WorkspaceState state = 1;
}

message GeneratedFile {
  // Path to the file relative to IdeAnalysis.build_artifact_root.
  string path = 1;

  // The text of the generated file, if not provided contents will be read
  // from the path above in user's workstation.
  optional bytes contents = 2;
}

message SourceFile {
  // Path to the source file relative to the repo root.
  string path = 1;

  // Working directory used by the build system. All the relative
  // paths in compiler_arguments should be relative to this path.
  // Relative to workspace root.
  string working_dir = 2;

  // Compiler arguments to compile the source file. If multiple variants
  // of the module being compiled are possible, the query script will choose
  // one.
  repeated string compiler_arguments = 3;

  // Any generated files that are used in compiling the file.
  repeated GeneratedFile generated = 4;

  // Paths to all of the sources, like build files, code generators,
  // proto files etc. that were used  during analysis. Used to figure
  // out when a set of build artifacts are stale and the query tool
  // must be re-run.
  // Relative to workspace root.
  repeated string deps = 5;

  // Represensts analysis status for this particular file. e.g. not part
  // of the build graph.
  optional Status status = 6;
}

message IdeAnalysis {
  // Path relative to workspace root, containing all the artifacts
  // generated by the build system. GeneratedFile.path are always
  // relative to this directory.
  string build_artifact_root = 1;

  repeated SourceFile sources = 2;

  // Status representing overall analysis.
  // Should fail only when no analysis can be performed, e.g. workspace
  // isn't setup.
  optional Status status = 3;
}
